util/election/runner_test.go

```diff
@@ -59,91 +59,101 @@ func TestConfigResignDelay(t *testing.T) {
 // TODO(pavelkalinnikov): Reduce flakiness risk in this test by making fewer
 // time assumptions.
 func TestElectionRunnerRun(t *testing.T) {
-	fakeTimeSource := clock.NewFake(time.Date(2016, 6, 28, 13, 40, 12, 45, time.UTC))
-	cfg := election.RunnerConfig{TimeSource: fakeTimeSource}
-	var tests = []struct {
+	const logID = "6962"
+	checkMaster := func(t *testing.T, held []string, wantMaster bool) {
+		t.Helper()
+		if got := (len(held) > 0 && held[0] == logID); got != wantMaster {
+			t.Errorf("holding: %v: master=%v, want %v", held, got, wantMaster)
+		}
+	}
+
+	for _, tc := range []struct {
 		desc       string
-		election   *stub.MasterElection
-		lostMaster bool
+		errs       to.Errs
+		isMaster   bool
 		wantMaster bool
+		loseMaster bool
+		resign     bool
 	}{
-		// Basic cases
-		{
-			desc:     "IsNotMaster",
-			election: stub.NewMasterElection(false, nil),
-		},
-		{
-			desc:       "IsMaster",
-			election:   stub.NewMasterElection(true, nil),
-			wantMaster: true,
-		},
-		{
-			desc:       "LostMaster",
-			election:   stub.NewMasterElection(true, nil),
-			lostMaster: true,
-			wantMaster: false,
-		},
-		// Error cases
-		{
-			desc: "ErrorOnStart",
-			election: stub.NewMasterElection(false,
-				&stub.Errors{Start: errors.New("on start")}),
-		},
-		{
-			desc: "ErrorOnWait",
-			election: stub.NewMasterElection(false,
-				&stub.Errors{Wait: errors.New("on wait")}),
-		},
+		// Basic cases.
+		{desc: "not-master"},
+		{desc: "is-master", isMaster: true, wantMaster: true},
+		{desc: "lose-master", isMaster: true, wantMaster: true, loseMaster: true},
+		{desc: "resign", isMaster: true, wantMaster: true, resign: true},
+		// Error cases.
+		{desc: "err-await", errs: to.Errs{Await: errors.New("ErrAwait")}},
+		{desc: "err-mctx", errs: to.Errs{WithMastership: errors.New("ErrMastership")}},
 		{
-			desc: "ErrorOnIsMaster",
-			election: stub.NewMasterElection(true,
-				&stub.Errors{IsMaster: errors.New("on IsMaster")}),
-			wantMaster: true,
+			desc:     "err-mctx-after-master",
+			isMaster: true,
+			errs:     to.Errs{WithMastership: errors.New("ErrMastership")},
 		},
+		{desc: "err-resign", errs: to.Errs{Resign: errors.New("ErrResign")}},
 		{
-			desc: "ErrorOnResign",
-			election: stub.NewMasterElection(true,
-				&stub.Errors{Resign: errors.New("resignation failure")}),
+			desc:       "err-resign-after-master",
+			isMaster:   true,
 			wantMaster: true,
+			errs:       to.Errs{Resign: errors.New("ErrResign")},
 		},
-	}
-	const logID = "6962"
-	for _, test := range tests {
-		t.Run(test.desc, func(t *testing.T) {
-			var wg sync.WaitGroup
+	} {
+		t.Run(tc.desc, func(t *testing.T) {
 			ctx, cancel := context.WithCancel(context.Background())
+			defer cancel()
 
-			startTime := time.Now()
-			fakeTimeSource := clock.NewFake(startTime)
+			e := to.NewElection()
+			d := to.NewDecorator(e)
```

Comment:
If you don't use e anywhere, you might be able to nest
`d := to.NewDecorator(to.NewElection())`

Reply:
It's used one time in line 134. I could use `d` there, but I would like that line to to ignore any `Resign` errors in the decorator, to simulate mastership loss properly.